  ## Task 3.Cuisine Classification


In [67]:
import numpy as np
import pandas as pd

In [68]:
# read csv file
df=pd.read_csv("12.Dataset.csv")
df.isnull().sum()

Restaurant ID           0
Restaurant Name         0
Country Code            0
City                    0
Address                 0
Locality                0
Locality Verbose        0
Longitude               0
Latitude                0
Cuisines                9
Average Cost for two    0
Currency                0
Has Table booking       0
Has Online delivery     0
Is delivering now       0
Switch to order menu    0
Price range             0
Aggregate rating        0
Rating color            0
Rating text             0
Votes                   0
dtype: int64

In [69]:
df.dropna(subset='Cuisines',inplace=True)

### 1.Encoding categorical varibles


In [70]:
#find frequency of each Cuisines values
fre=df['Cuisines'].value_counts()
#map frequency in the dataset
df['Cuisinesen']=df['Cuisines'].map(fre)

In [72]:
df.head(2)

Restaurant ID   Restaurant Name  Country Code         City  \
0        6317637  Le Petit Souffle           162  Makati City   
1        6304287  Izakaya Kikufuji           162  Makati City   

                                             Address  \
0  Third Floor, Century City Mall, Kalayaan Avenu...   
1  Little Tokyo, 2277 Chino Roces Avenue, Legaspi...   

                                     Locality  \
0   Century City Mall, Poblacion, Makati City   
1  Little Tokyo, Legaspi Village, Makati City   

                                    Locality Verbose   Longitude   Latitude  \
0  Century City Mall, Poblacion, Makati City, Mak...  121.027535  14.565443   
1  Little Tokyo, Legaspi Village, Makati City, Ma...  121.014101  14.553708   

                     Cuisines  ...  Has Table booking Has Online delivery  \
0  French, Japanese, Desserts  ...                Yes                  No   
1                    Japanese  ...                Yes                  No   

  Is delivering now Switch to order menu Price range Aggregate rating  \
0                No                   No           3              4.8   
1                No                   No           3              4.5   

   Rating color  Rating text Votes Cuisinesen  
0    Dark Green    Excellent   314          1  
1    Dark Green    Excellent   591         17  

[2 rows x 22 columns]

In [73]:
#split object and float columns
ob=[i for i in df.columns if df[i].dtypes=="O"]
it=[i for i in df.columns if df[i].dtypes!="O"]

In [74]:
#find object unique values
for i in ob:
    a=df[i].unique()
    print(i)
    print(a)
    print()

Restaurant Name
['Le Petit Souffle' 'Izakaya Kikufuji' 'Heat - Edsa Shangri-La' ...
 'Huqqa' 'A���k Kahve' "Walter's Coffee Roastery"]

City
['Makati City' 'Mandaluyong City' 'Pasay City' 'Pasig City' 'Quezon City'
 'San Juan City' 'Santa Rosa' 'Tagaytay City' 'Taguig City' 'Bras�_lia'
 'Rio de Janeiro' 'S��o Paulo' 'Albany' 'Armidale' 'Athens' 'Augusta'
 'Balingup' 'Beechworth' 'Boise' 'Cedar Rapids/Iowa City' 'Chatham-Kent'
 'Clatskanie' 'Cochrane' 'Columbus' 'Consort' 'Dalton' 'Davenport'
 'Des Moines' 'Dicky Beach' 'Dubuque' 'East Ballina' 'Fernley' 'Flaxton'
 'Forrest' 'Gainesville' 'Hepburn Springs' 'Huskisson' 'Inverloch'
 'Lakes Entrance' 'Lakeview' 'Lincoln' 'Lorn' 'Macedon' 'Macon' 'Mayfield'
 'Mc Millan' 'Middleton Beach' 'Monroe' 'Montville' 'Ojo Caliente'
 'Orlando' 'Palm Cove' 'Paynesville' 'Penola' 'Pensacola' 'Phillip Island'
 'Pocatello' 'Potrero' 'Princeton' 'Rest of Hawaii' 'Savannah' 'Singapore'
 'Sioux City' 'Tampa Bay' 'Tanunda' 'Trentham East' 'Valdosta' 'Vernoni

In [75]:
# split feature with more and small number of unique value
df[ob].columns

Index(['Restaurant Name', 'City', 'Address', 'Locality', 'Locality Verbose',
       'Cuisines', 'Currency', 'Has Table booking', 'Has Online delivery',
       'Is delivering now', 'Switch to order menu', 'Rating color',
       'Rating text'],
      dtype='object')

In [76]:
la=['Has Table booking', 'Has Online delivery',
       'Is delivering now', 'Switch to order menu', 'Rating color',
       'Rating text']
ta=['Restaurant Name', 'City', 'Address', 'Locality', 'Locality Verbose',
       'Cuisines', 'Currency']

In [77]:
from sklearn.preprocessing import LabelEncoder
import category_encoders as ce
L=LabelEncoder()
T=ce.TargetEncoder()
for i in la:
    df[i]=L.fit_transform(df[i])
for i in ta:
    df[i]=T.fit_transform(df[i],df['Cuisinesen'])

In [78]:
df

Restaurant ID  Restaurant Name  Country Code        City     Address  \
0           6317637       176.162896           162  174.933196  176.162896   
1           6304287       178.244632           162  174.933196  178.244632   
2           6300002       176.162896           162  184.158984  176.162896   
3           6318506       178.374741           162  184.158984  178.374741   
4           6314302       176.423113           162  184.158984  176.423113   
...             ...              ...           ...         ...         ...   
9546        5915730       176.553222           208  155.131177  176.553222   
9547        5908749       176.162896           208  155.131177  176.162896   
9548        5915807       176.162896           208  155.131177  176.162896   
9549        5916112       176.293005           208  155.131177  176.293005   
9550        5927402       214.935222           208  155.131177  214.935222   

        Locality  Locality Verbose   Longitude   Latitude    Cuisines  ...  \
0     176.162896        176.162896  121.027535  14.565443  176.162896  ...   
1     178.244632        178.244632  121.014101  14.553708  123.479681  ...   
2     176.162896        176.162896  121.056831  14.581404  176.162896  ...   
3     190.412063        190.412063  121.056475  14.585318  119.368368  ...   
4     190.412063        190.412063  121.057508  14.584450  171.567301  ...   
...          ...               ...         ...        ...         ...  ...   
9546  181.177310        181.177310   28.977392  41.022793  169.040639  ...   
9547  176.162896        176.162896   29.041297  41.009847  176.162896  ...   
9548  173.869313        173.869313   29.034640  41.055817  176.162896  ...   
9549  173.869313        173.869313   29.036019  41.057979  173.940239  ...   
9550  214.935222        214.935222   29.026016  40.984776  299.000000  ...   

      Has Table booking  Has Online delivery  Is delivering now  \
0                     1                    0                  0   
1                     1                    0                  0   
2                     1                    0                  0   
3                     0                    0                  0   
4                     1                    0                  0   
...                 ...                  ...                ...   
9546                  0                    0                  0   
9547                  0                    0                  0   
9548                  0                    0                  0   
9549                  0                    0                  0   
9550                  0                    0                  0   

      Switch to order menu  Price range  Aggregate rating  Rating color  \
0                        0            3               4.8             0   
1                        0            3               4.5             0   
2                        0            4               4.4             1   
3                        0            4               4.9             0   
4                        0            4               4.8             0   
...                    ...          ...               ...           ...   
9546                     0            3               4.1             1   
9547                     0            3               4.2             1   
9548                     0            4               3.7             5   
9549                     0            4               4.0             1   
9550                     0            2               4.0             1   

      Rating text  Votes  Cuisinesen  
0               1    314           1  
1               1    591          17  
2               5    270           1  
3               1    365          18  
4               1    229           3  
...           ...    ...         ...  
9546            5    788           4  
9547            5   1034           1  
9548            2    661           1  
9549            5    901           2

In [79]:
# find correlation between dependent and indepentend variable
c=abs(df.corr()['Cuisinesen'].sort_values(ascending=False))

In [80]:
c

Cuisinesen              1.000000
Address                 0.974326
Cuisines                0.966443
Restaurant Name         0.843902
Locality Verbose        0.364289
Locality                0.359053
City                    0.242290
Currency                0.201446
Longitude               0.131914
Rating color            0.060251
Latitude                0.046382
Restaurant ID           0.003971
Is delivering now       0.010331
Average Cost for two    0.031630
Has Online delivery     0.072240
Rating text             0.081568
Has Table booking       0.114553
Votes                   0.146203
Country Code            0.192742
Price range             0.245731
Aggregate rating        0.279136
Switch to order menu         NaN
Name: Cuisinesen, dtype: float64

In [81]:
#drop porr correlation
c.index

Index(['Cuisinesen', 'Address', 'Cuisines', 'Restaurant Name',
       'Locality Verbose', 'Locality', 'City', 'Currency', 'Longitude',
       'Rating color', 'Latitude', 'Restaurant ID', 'Is delivering now',
       'Average Cost for two', 'Has Online delivery', 'Rating text',
       'Has Table booking', 'Votes', 'Country Code', 'Price range',
       'Aggregate rating', 'Switch to order menu'],
      dtype='object')

In [82]:
dro=[ 'Restaurant ID', 'Is delivering now',
       'Average Cost for two', 'Has Online delivery', 'Rating text',
       'Has Table booking', 'Votes', 'Country Code', 'Price range',
       'Aggregate rating', 'Switch to order menu']
df.drop(dro,axis=1,inplace=True)

### 2.split data into training and testing


In [83]:
#split feature and label
X=df.drop('Cuisinesen',axis=1)
y=df['Cuisinesen']

In [84]:
#split train and test
from sklearn.model_selection import train_test_split
X_train,X_test,y_tarin,y_test=train_test_split(X,y,test_size=0.25,random_state=45)

In [85]:
#scaling feature
from sklearn.preprocessing import StandardScaler
s=StandardScaler()
X_train=s.fit_transform(X_train)
X_test=s.fit_transform(X_test)

### 3.Build the model

In [88]:
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import precision_score,recall_score,accuracy_score
m=[LogisticRegression(max_iter=1000),RandomForestClassifier()]
for i in m:
    i.fit(X_train,y_train)
    pred=i.predict(X_test)
    pre=precision_score(pred,y_test,average='macro',zero_division=0)
    rec=recall_score(pred,y_test,average='macro',zero_division=0)
    acc=accuracy_score(pred,y_test)
    print(i)
    print('precision_score : ',pre)
    print('recall_score : ',rec)
    print('accuracy_score : ',acc)
    print()


LogisticRegression(max_iter=1000)
precision_score :  0.16773263106753927
recall_score :  0.16595248392129183
accuracy_score :  0.4572506286672255

RandomForestClassifier()
precision_score :  0.48564302734074255
recall_score :  0.5307864612014315
accuracy_score :  0.5984911986588433



i got randomforest 59% and logistic 45% accuracy ,althrough random forest better than logistic but it is not good ,i will use cross validation for increasing model accuracy

### 4.Build cross validation

In [89]:
from sklearn.model_selection import cross_validate
from sklearn.pipeline import make_pipeline
from sklearn.metrics import make_scorer

rf=RandomForestClassifier()
pipe=make_pipeline(StandardScaler(),rf)

cv=cross_validate(pipe,X_train,y_train,cv=5,
                 scoring={'acc':'accuracy',
                         'precision':make_scorer(precision_score,average='macro',zero_division=0),
                         'recall':make_scorer(recall_score,average='macro',zero_division=0)},
                 return_train_score=False)
print('cross_validate')
print(f"Accuracy: {cv['test_acc'].mean():.4f}±{cv['test_acc'].std():4f}")
print(f"Precision: {cv['test_precision'].mean():.4f}±{cv['test_precision'].std():4f}")
print(f"Recall: {cv['test_recall'].mean():.4f}±{cv['test_recall'].std():4f}")

cross_validate
Accuracy: 0.9561±0.002777
Precision: 0.9090±0.017884
Recall: 0.8782±0.007998


finaly i got 95% accuracy ,precision is 90.9% and recall is 87.8% its a good model and data training good model perform very well the standard deviation is good

# Thank you